bloom.go

```diff
@@ -0,0 +1,890 @@
+const bloomSizeBase = 12252240
+
+// empirically determined to achieve 1% FPR
+var bloomDefaultHash = bloomHasherCRCBlocked64B8K3
+
+const bloomDefaultLoad = 0.42
+
+// Castagnoli CRCs have hardware instructions to compute them.
+var crcTab = crc32.MakeTable(crc32.Castagnoli)
+
+func makeBloomFilterEmpty() bloom {
+	bloomWordOnce.Do(initBloomWordTab)
+	return bloom{bloomDefaultHash, make([]uint8, bloomSizeBase)}
+}
+
+func makeBloomFilterWithHasher(hash bloomHash) bloom {
+	bloomWordOnce.Do(initBloomWordTab)
+	return bloom{hash, make([]uint8, bloomSizeBase)}
+}
+
+func (b *bloom) Len() int {
+	return len(b.bits) * 8
+}
+
+func (b *bloom) add(xs []uint32) {
+	for _, x := range xs {
+		b.bits[int(x/8)%len(b.bits)] |= 1 << (x % 8)
+	}
+}
+
+func (b *bloom) addBytes(data []byte) {
+	b.add(b.hasher(data))
+}
+
+func (b *bloom) maybeHas(xs []uint32) bool {
+	for _, x := range xs {
+		if b.bits[int(x/8)%len(b.bits)]&(1<<(x%8)) == 0 {
+			return false
+		}
+	}
+	return true
+}
+
+func (b *bloom) maybeHasBytes(xs []byte) bool {
+	return b.maybeHas(b.hasher(xs))
+}
+
+func (b *bloom) load() float64 {
+	total := 0
+	for _, x := range b.bits {
+		total += bits.OnesCount8(x)
+	}
+	return float64(total) / float64(len(b.bits)*8)
+}
+
+// shrinkToSize returns a resized bloom filter with a bit density close to target.
+// This exploits the fact that a test for a probe x in the bloom filter is actually
+// a test for bit x%len, and a bloom filter of size newlen that divides len is easily
+// derived by oring the bits together len/newlen times. This works because
+// x%newlen == x%len%newlen iff newlen divides len, so we can shrink the bloom filter
+// without having the original probes or keys! This functionality lets us construct
+// a bloom filter while only having an upper bound on cardinality, instead of having
+// to have a separate, expensive input-counting phase.
+func (b *bloom) shrinkToSize(target float64) bloom {
+	if target <= 0.0 || target >= 1.0 {
+		return *b
+	}
+
+	// shrinking sets each output bit to the OR-ed together
+	// output of k=`factor` bits that are set with probability
+	// x=`b.load()`. We want to achieve a target load `y`.
+	//
+	// The probability that a bit is set is one minus the probability
+	// that its inputs are all unset-- 1-(1-x)^k. To get k given y,
+	// https://www.wolframalpha.com/input/?i=solve+for+k+in+y%3D1-%281-x%29%5Ek
+	// => k=log(1-y)/log(1-x)
+	factor := int(math.Log(1-target) / math.Log(1-b.load()))
+
+	// We can only shrink the bloom filter to a size
+	// that is a factor of the input size. This is made easier by bloomSizeBase
+	// being highly composite.
+	for factor > 0 && len(b.bits)%factor != 0 {
+		factor--
+	}
+
+	if factor <= 1 {
+		return *b
+	}
+	out := bloom{b.hasher, make([]uint8, len(b.bits)/factor)}
+	j := 0
+	for i := 0; i < len(b.bits); i++ {
+		out.bits[j] |= b.bits[i]
+		j++
+		if j >= len(out.bits) {
+			j = 0
+		}
+	}
+
+	return out
+}
+
+func (b *bloom) GobEncode() ([]byte, error) {
```

Comment:
I think you can implement binary marshalling interface to avoid gob?

Reply:
I will drop the Gob interface when bumping the zoekt file version. This was just the first encoding interface that I saw.

MarshalBinary / UnmarshalBinary exactly equivalent interfaces, anyways.